pkg/decoder/delimiter.go

```diff
@@ -0,0 +1,176 @@
+	}
+}
+
+// Delimiter receives Lines to form content
+type Delimiter interface {
+	LineChan() chan<- *Line
+	Start()
+	Stop()
+}
+
+// SingleLineDelimiter reads lines from LineChan and uses lineBuffer to send them
+type SingleLineDelimiter struct {
+	lineChan   chan *Line
+	lineBuffer *LineBuffer
+}
+
+// NewSingleLineDelimiter returns a new NewSingleLineDelimiter
+func NewSingleLineDelimiter(outputChan chan *Output) *SingleLineDelimiter {
+	lineChan := make(chan *Line)
+	lineBuffer := NewLineBuffer(outputChan)
+	delimiter := SingleLineDelimiter{
+		lineChan:   lineChan,
+		lineBuffer: lineBuffer,
+	}
+	return &delimiter
+}
+
+// LineChan returns lineChan
+func (d *SingleLineDelimiter) LineChan() chan<- *Line {
+	return d.lineChan
+}
+
+// Start starts the delimiter
+func (d *SingleLineDelimiter) Start() {
+	go d.run()
+}
+
+// Start stops the delimiter
+func (d *SingleLineDelimiter) Stop() {
+	close(d.lineChan)
+}
+
+// run reads lines from lineChan to process them
+func (d *SingleLineDelimiter) run() {
+	for line := range d.lineChan {
+		d.process(line)
+	}
+}
+
+// process sends or truncates lines using lineBuffer
+func (d *SingleLineDelimiter) process(line *Line) {
+	lineLen := len(line.content)
+	if lineLen < contentLenLimit {
+		// send line
+		d.lineBuffer.Add(line)
+		d.lineBuffer.Flush()
+	} else {
+		// truncate content
+		d.lineBuffer.AddAndTruncate(line)
+	}
+}
+
+// outdatedDuration represents the time we want to wait before sending content stored in lineBuffer
+// when no more line is received
+const outdatedDuration = 1 * time.Second
+
+// MultiLineDelimiter reads lines from lineChan and uses lineBuffer to send them when a new line matches with re or the timer is fired
+type MultiLineDelimiter struct {
+	lineChan     chan *Line
+	lineBuffer   *LineBuffer
+	newContentRe *regexp.Regexp
+	timer        *time.Timer
+	mu           sync.Mutex
+	shouldStop   bool
+}
+
+// NewMultiLineDelimiter returns a new MultiLineDelimiter
+func NewMultiLineDelimiter(outputChan chan *Output, newContentRe *regexp.Regexp) *MultiLineDelimiter {
+	lineChan := make(chan *Line)
+	lineBuffer := NewLineBuffer(outputChan)
+	timer := time.NewTimer(outdatedDuration)
+	delimiter := MultiLineDelimiter{
+		lineChan:     lineChan,
+		lineBuffer:   lineBuffer,
+		newContentRe: newContentRe,
+		timer:        timer,
+	}
+	return &delimiter
+}
+
+// LineChan returns lineChan
+func (d *MultiLineDelimiter) LineChan() chan<- *Line {
+	return d.lineChan
+}
+
+// Start starts the delimiter
+func (d *MultiLineDelimiter) Start() {
+	go d.run()
+}
+
+// Start stops the delimiter
+func (d *MultiLineDelimiter) Stop() {
+	d.mu.Lock()
+	close(d.lineChan)
+	d.shouldStop = true
+	// assure to stop timer goroutine
+	d.timer.Reset(outdatedDuration)
+	d.mu.Unlock()
+}
+
+// run sends content when timer expires and reads lines from lineChan to process them
+func (d *MultiLineDelimiter) run() {
+	// send content safely when timer fires
+	go func() {
+		for range d.timer.C {
```

Comment:
what about moving this to a method? and `go d.method()` it in `Start()`?

Reply:
I like it here because we understand why we need to use a `mutex`, WDYT ?

Reply:
I think it should be a proper method, have a name, it will be easier to understand